smqtk_descriptors/impls/descriptor_generator/pytorch.py

```diff
@@ -0,0 +1,538 @@
+        us enabled.
+    :param None|int|float|str normalize:
+        Optionally normalize descriptor vectors as they are produced. We use
+        ``numpy.linalg.norm`` so any valid value for the ``ord`` parameter is
+        acceptable here.
+    :param bool iter_runtime:
+        By default the input image matrix data is accumulated into a list in
+        order to utilize `torch.utils.data.DataLoader` for batching.
+        If this is parameter is True however, we use a custom parallel
+        iteration pipeline for image transformation into network appropriate
+        tensors.
+        Using this mode changes the maximum RAM usage profile to be linear with
+        batch-size instead of input image data amount, as well as having lower
+        latency to first yield.
+        This mode is idea for streaming input or high input volume situations.
+        This mode currently has a short-coming of working only in a threaded
+        manner so it is not as fast as using the `DataLoader` avenue.
+    :param bool global_average_pool:
+        Optionally apply a GAP operation to the spatial dimension of the feature
+        vector that is returned by the descriptor generator. Some models return
+        a w x h x k tensor and flatten them into a single dimension, which can
+        remove the spatial context of the features. Taking an average over each
+        channel can improve the clustering in some cases.
+    """
+
+    @classmethod
+    def is_usable(cls) -> bool:
+        valid = torch is not None
+        if not valid:
+            LOG.debug("Torch python module not imported")
+        return valid
+
+    @classmethod
+    def get_default_config(cls) -> Dict[str, Any]:
+        c = super().get_default_config()
+        c['image_reader'] = make_default_config(ImageReader.get_impls())
+        return c
+
+    @classmethod
+    def from_config(
+        cls: Type[T],
+        config_dict: Dict,
+        merge_default: bool = True
+    ) -> T:
+        # Copy config to prevent input modification
+        config_dict = copy.deepcopy(config_dict)
+        config_dict['image_reader'] = \
+            from_config_dict(config_dict['image_reader'],
+                             ImageReader.get_impls())
+        return super().from_config(config_dict, merge_default)
+
+    def __init__(
+        self,
+        image_reader: ImageReader,
+        image_load_threads: Optional[int] = 1,
+        weights_filepath: Optional[str] = None,
+        image_tform_threads: Optional[int] = 1,
+        batch_size: int = 32,
+        use_gpu: bool = False,
+        cuda_device: Optional[int] = None,
+        normalize: Optional[Union[int, float, str]] = None,
+        iter_runtime: bool = False,
+        global_average_pool: bool = False
+    ):
+        super().__init__()
+
+        # Make sure given image reader is valid
+        try:
+            image_reader.is_valid_element
+        except AttributeError:
+            raise
```

Comment:
I'm not sure I understand what this block is for. I don't see it in the source we are drawing from, and I don't know what this is doing: the except catches an exception, does nothing with it and then just re-raises it. I either suggest we don't need this block or this needs more explanation comments on why this is here.

Reply:
I had added it to make sure that the provided image_reader parameter is valid. Tested in `test_init_no_image_reader`, which is based on the `test_init_no_*` tests in test_caffe.py. I went ahead and changed it to `assert image_reader is not None`, so hopefully it's a bit more clear now.